### PR TITLE
Add operator kwarg to Cols

### PR DESCRIPTION
**Pull request by bkamins**
x-ref https://github.com/JuliaData/DataFrames.jl/pull/3224

### CODE REVIEW

**Comment by bkamins** [comment]
CC @jar @pdeffebach @yjunechoe

File: src/DataAPI.jl [hunk 1]
```diff
@@ -183,9 +183,9 @@ Between(x::Union{Int, Symbol}, y::AbstractString) = Between(x, Symbol(y))
 Between(x::AbstractString, y::Union{Int, Symbol}) = Between(Symbol(x), y)
 
 """
```
Comment (bkamins) [inline src/DataAPI.jl:185]: this change removes long deprecated old behavior description.

File: src/DataAPI.jl [hunk 2]
```diff
@@ -183,9 +183,9 @@ Between(x::Union{Int, Symbol}, y::AbstractString) = Between(x, Symbol(y))
...
-    All(cols...)
+    All()
 
-Select the union of the selections in `cols`. If `cols == ()`, select all columns.
+Select all columns.
```
Comment (bkamins) [inline src/DataAPI.jl:188]: this change removes long deprecated old behavior description.

File: src/DataAPI.jl [hunk 3]
```diff
@@ -198,17 +198,27 @@ struct All{T<:Tuple}
 end
 
 """
-    Cols(cols...)
-    Cols(f::Function)
+    Cols(cols...; operation::Symbol=:union)
+    Cols(f::Function; operation::Symbol=:union)
 
-Select the union of the selections in `cols`. If `cols == ()`, select no columns.
+Select a union of the selections in `cols`. If `cols == ()`, select no columns.
```
Comment (nalimilan) [inline src/DataAPI.jl:204]:
```suggestion
Select columns matching specifications in `cols`. If `cols == ()`, select no columns.
```

File: src/DataAPI.jl [hunk 4]
```diff
@@ -198,17 +198,27 @@ struct All{T<:Tuple}
...
 
 If the only positional argument is a `Function` `f` then select the columns whose
 names passed to the `f` predicate as strings return `true`.
+
+If `operation` keyword argument is passed then it can be either `:union` (the default)
+or `:intersect`. If it is `:intersect` then an intersection of the selections
+in `cols` is selected.
```
Comment (nalimilan) [inline src/DataAPI.jl:211]:
```suggestion
When multiple `cols` selectors are passed,  if `operation=:union` (the default)
all columns matching at least one selector are returned, and if
`operation=:intersect` only columns matching all selectors are returned.
```

**Review by nalimilan — APPROVED** [review]

**Comment by adienes** [comment]
I think we discussed this in the previous issue but I forgot---could you remind me why we use keywords here rather than a function? Then it would have `Cols(...; operation=any)` and `Cols(...; operation=all)` using built in functions `any` and `all`, and could use user-defined predicates.

**Comment by bkamins** [comment]
We have not discussed it previously. Actually what you propose is a good idea. But `any` and `all` are not correct functions to be passed. We can say that `union` function is the default and `intersect` is a function to pass if one wants intersection. However, potentially user could pass any function operating on sets and returning a set as a result.

Then the definition in DataFrames.jl would be:
```
@inline Base.getindex(x::AbstractIndex, idx::Cols) =
    isempty(idx.cols) ? Int[] : idx.operation(getindex.(Ref(x), idx.cols)...)
```
@nalimilan - what do you think?

**Comment by adienes** [comment]
Makes sense. Then we'd also get `setdiff` and `symdiff` for free

**Comment by jariji** [comment]
Yep, I agree with adienes.

**Comment by nalimilan** [comment]
Yeah why not. Maybe we need a more specific name than "operation" then for the argument. Again the same discussion as the one about "combine" in `unstack`? :-)

**Comment by bkamins** [comment]
AFAICT this fits well. In mathematics `union` and `intersect` etc. are typically called "operations", c.f. e.g. https://en.wikipedia.org/wiki/Set_(mathematics)#Basic_operations

File: src/DataAPI.jl [hunk 5]
```diff
@@ -198,17 +198,27 @@ struct All{T<:Tuple}
 end
 
 """
-    Cols(cols...)
-    Cols(f::Function)
+    Cols(cols...; operation::Symbol=:union)
+    Cols(f::Function; operation::Symbol=:union)
```
Comment (bkamins) [inline src/DataAPI.jl:202]:
```suggestion
    Cols(cols...; operation=union)
    Cols(f::Function; operation=union)
```

File: src/DataAPI.jl [hunk 6]
```diff
@@ -198,17 +198,27 @@ struct All{T<:Tuple}
...
 
-Select the union of the selections in `cols`. If `cols == ()`, select no columns.
+Select columns matching specifications in `cols`. If `cols == ()`, select no columns.
 
 If the only positional argument is a `Function` `f` then select the columns whose
 names passed to the `f` predicate as strings return `true`.
+
+When multiple `cols` selectors are passed,  if `operation=:union` (the default)
+all columns matching at least one selector are returned, and if
+`operation=:intersect` only columns matching all selectors are returned.
```
Comment (bkamins) [inline src/DataAPI.jl:211]:
```suggestion
When multiple `cols` selectors are passed then the sets of columns selected by them
are passed to `operation` callable as positional arguments.
`operation` should be a set operation function, like `union`, `intersect`, `setdiff`, and `symdiff`
defined in Base Julia. By default `operation=union` in which case all columns matching
at least one selector are returned.
```

File: src/DataAPI.jl [hunk 7]
```diff
@@ -198,17 +198,27 @@ struct All{T<:Tuple}
...
 struct Cols{T<:Tuple}
     cols::T
-    Cols(args...) = new{typeof(args)}(args)
+    operation::Symbol
+    function Cols(args...; operation::Symbol=:union)
+        if !(operation == :union || operation == :intersect)
+            throw(ArgumentError("operation must be `:union` or `:intersect`"))
+        end
+        return new{typeof(args)}(args, operation)
+    end
```
Comment (bkamins) [inline src/DataAPI.jl:221]:
```suggestion
    operation
    Cols(args...; operation::Symbol=union) = new{typeof(args)}(args, operation)
    end
```

File: test/runtests.jl [hunk 8]
```diff
@@ -249,6 +249,9 @@ end
             (Ref(DataAPI.BroadcastedSelector(v())) .=> [sum, float])
     end
 
+    @test DataAPI.Cols(:a, operation=:union).operation == :union
+    @test DataAPI.Cols(:a, operation=:intersect).operation == :intersect
+    @test_throws ArgumentError DataAPI.Cols(:a, operation=:bad)
```
Comment (bkamins) [inline test/runtests.jl:254]:
```suggestion
    @test DataAPI.Cols(:a, operation=union).operation == union
    @test DataAPI.Cols(:a, operation=intersect).operation == intersect
```

File: src/DataAPI.jl [hunk 9]
```diff
@@ -198,17 +198,25 @@ struct All{T<:Tuple}
...
+`operation` should be a set operation function, like `union`, `intersect`, `setdiff`, and `symdiff`
+defined in Base Julia. By default `operation=union` in which case all columns matching
+at least one selector are returned.
 """
 struct Cols{T<:Tuple}
     cols::T
-    Cols(args...) = new{typeof(args)}(args)
+    operation
+    Cols(args...; operation::Symbol=union) = new{typeof(args)}(args, operation)
+    end
```
Comment (bkamins) [inline src/DataAPI.jl:219]:
```suggestion
```

File: src/DataAPI.jl [hunk 10]
```diff
@@ -198,17 +198,24 @@ struct All{T<:Tuple}
...
+are passed to `operation` callable as positional arguments.
+`operation` should be a set operation function, like `union`, `intersect`, `setdiff`, and `symdiff`
+defined in Base Julia. By default `operation=union` in which case all columns matching
+at least one selector are returned.
 """
 struct Cols{T<:Tuple}
     cols::T
-    Cols(args...) = new{typeof(args)}(args)
+    operation
+    Cols(args...; operation::Symbol=union) = new{typeof(args)}(args, operation)
```
Comment (bkamins) [inline src/DataAPI.jl:218]:
```suggestion
    Cols(args...; operation=union) = new{typeof(args)}(args, operation)
```

**Comment by bkamins** [comment]
I have updated the PR.

**Comment by jariji** [comment]
The tests don't actually test the functionality.

**Comment by adienes** [comment]
Perhaps `operator` is very slightly a better choice for some very tenuous reasons:

1. often the arguments in `Cols` will be indicator functions of sets, and a function that acts on other functions is often called an operator
2. to me it scans slightly more clearly that what follows `operator=` should be a verb (i.e. a function)
3. it is 1 character shorter :)

The wikipedia page for the [Algebra of Sets](https://en.wikipedia.org/wiki/Algebra_of_sets) uses both terms interchangeably, so I'm not sure which is more standard.

**Comment by bkamins** [comment]
I checked several references, and the border between `operation` and `operator` is thin. [Operation (mathematics)](https://en.wikipedia.org/wiki/Operation_(mathematics)) gives a most clear distinction:

> An operator is similar to an operation in that it refers to the symbol or the process used to denote the operation, hence their point of view is different. For instance, one often speaks of "the operation of addition" or "the addition operation", when focusing on the operands and result, but one switches to "addition operator" (rarely "operator of addition"), when focusing on the process

Given this explanation indeed `operator` seems better, but I am not sure.

File: src/DataAPI.jl [hunk 11]
```diff
@@ -198,17 +198,24 @@ struct All{T<:Tuple}
 end
 
 """
-    Cols(cols...)
-    Cols(f::Function)
+    Cols(cols...; operation=union)
+    Cols(f::Function; operation=union)
```
Comment (bkamins) [inline src/DataAPI.jl:202]:
```suggestion
    Cols(cols...; operator=union)
    Cols(f::Function; operator=union)
```

File: src/DataAPI.jl [hunk 12]
```diff
@@ -198,17 +198,24 @@ struct All{T<:Tuple}
...
-Select the union of the selections in `cols`. If `cols == ()`, select no columns.
+Select columns matching specifications in `cols`. If `cols == ()`, select no columns.
 
 If the only positional argument is a `Function` `f` then select the columns whose
 names passed to the `f` predicate as strings return `true`.
+
+When multiple `cols` selectors are passed then the sets of columns selected by them
+are passed to `operation` callable as positional arguments.
+`operation` should be a set operation function, like `union`, `intersect`, `setdiff`, and `symdiff`
+defined in Base Julia. By default `operation=union` in which case all columns matching
```
Comment (bkamins) [inline src/DataAPI.jl:212]:
```suggestion
are passed to `operator` callable as positional arguments.
`operator` should be a set operation function, like `union`, `intersect`, `setdiff`, and `symdiff`
defined in Base Julia. By default `operator=union` in which case all columns matching
```

File: src/DataAPI.jl [hunk 13]
```diff
@@ -198,17 +198,24 @@ struct All{T<:Tuple}
...
+at least one selector are returned.
 """
 struct Cols{T<:Tuple}
     cols::T
-    Cols(args...) = new{typeof(args)}(args)
+    operation
+    Cols(args...; operation=union) = new{typeof(args)}(args, operation)
```
Comment (bkamins) [inline src/DataAPI.jl:218]:
```suggestion
    operator
    Cols(args...; operator=union) = new{typeof(args)}(args, operator)
```

File: test/runtests.jl [hunk 14]
```diff
@@ -249,6 +249,8 @@ end
             (Ref(DataAPI.BroadcastedSelector(v())) .=> [sum, float])
     end
 
+    @test DataAPI.Cols(:a, operation=union).operation == union
+    @test DataAPI.Cols(:a, operation=intersect).operation == intersect
```
Comment (bkamins) [inline test/runtests.jl:253]:
```suggestion
    @test DataAPI.Cols(:a, operator=union).operator == union
    @test DataAPI.Cols(:a, operator=intersect).operator == intersect
```

File: src/DataAPI.jl [hunk 15]
```diff
@@ -198,17 +198,24 @@ struct All{T<:Tuple}
...
+    Cols(f::Function; operator=union)
 
-Select the union of the selections in `cols`. If `cols == ()`, select no columns.
+Select columns matching specifications in `cols`. If `cols == ()`, select no columns.
 
 If the only positional argument is a `Function` `f` then select the columns whose
 names passed to the `f` predicate as strings return `true`.
+
+When multiple `cols` selectors are passed then the sets of columns selected by them
+are passed to `operator` callable as positional arguments.
```
Comment (nalimilan) [inline src/DataAPI.jl:210]:
```suggestion
are passed to `operator` as positional arguments.
```
(or "the `operator` callable", but the next sentence says it must be a function)

**Review by nalimilan — APPROVED** [review]

**Comment by bkamins** [comment]
Thank you!

**Pull request merged**